plugins/outputs/prometheus_remote_write/prometheus_remote_write.go

```diff
@@ -0,0 +1,163 @@
+	URL           string `toml:"url"`
+	BasicUsername string `toml:"basic_username"`
+	BasicPassword string `toml:"basic_password"`
+	tls.ClientConfig
+
+	client http.Client
+}
+
+var sampleConfig = `
+  ## URL to send Prometheus remote write requests to.
+  url = "http://localhost/push"
+
+  ## Optional HTTP asic auth credentials.
+  # basic_username = "username"
+  # basic_password = "pa55w0rd"
+
+  ## Optional TLS Config for use on HTTP connections.
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use TLS but skip chain & host verification
+  # insecure_skip_verify = false
+`
+
+func (p *PrometheusRemoteWrite) Connect() error {
+	tlsConfig, err := p.ClientConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	p.client = http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+		},
+	}
+	return nil
+}
+
+func (p *PrometheusRemoteWrite) Close() error {
+	return nil
+}
+
+func (p *PrometheusRemoteWrite) Description() string {
+	return "Configuration for the Prometheus remote write client to spawn"
+}
+
+func (p *PrometheusRemoteWrite) SampleConfig() string {
+	return sampleConfig
+}
+
+func (p *PrometheusRemoteWrite) Write(metrics []telegraf.Metric) error {
+	var req prompb.WriteRequest
+
+	for _, metric := range metrics {
+		tags := metric.TagList()
+		commonLabels := make([]prompb.Label, 0, len(tags))
+		for _, tag := range tags {
+			commonLabels = append(commonLabels, prompb.Label{
+				Name:  prometheus_client.Sanitize(tag.Key),
+				Value: tag.Value,
+			})
+		}
+
+		for _, field := range metric.FieldList() {
+			labels := make([]prompb.Label, len(commonLabels), len(commonLabels)+1)
+			copy(labels, commonLabels)
+			labels = append(labels, prompb.Label{
+				Name:  "__name__",
+				Value: metric.Name() + "_" + field.Key,
+			})
+			sort.Sort(byName(labels))
+
+			// Ignore histograms and summaries.
+			switch metric.Type() {
+			case telegraf.Histogram, telegraf.Summary:
+				continue
+			}
+
+			// Ignore string and bool fields.
+			var value float64
+			switch fv := field.Value.(type) {
+			case int64:
+				value = float64(fv)
+			case uint64:
+				value = float64(fv)
+			case float64:
+				value = fv
+			default:
+				continue
+			}
+
+			req.Timeseries = append(req.Timeseries, prompb.TimeSeries{
+				Labels: labels,
+				Samples: []prompb.Sample{{
+					Timestamp: metric.Time().UnixNano() / int64(time.Millisecond),
+					Value:     value,
+				}},
+			})
+		}
+	}
+
+	buf, err := proto.Marshal(&req)
+	if err != nil {
+		return err
+	}
+
+	compressed := snappy.Encode(nil, buf)
+	httpReq, err := http.NewRequest("POST", p.URL, bytes.NewReader(compressed))
+	if err != nil {
+		return err
+	}
+	httpReq.Header.Add("Content-Encoding", "snappy")
+	httpReq.Header.Set("Content-Type", "application/x-protobuf")
+	httpReq.Header.Set("X-Prometheus-Remote-Write-Version", "0.1.0")
+	httpReq.Header.Set("User-Agent", "Telegraf/"+internal.Version())
```

Comment:
We should support custom headers. A use case scenario can be to separate tenants, for example cortex handles tenants by headers in http requests.